### PR TITLE
loader: add lock for memory cached checkpoint

### LINKED ISSUE #1143: "concurrent map read and map write" for loader checkpoint
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do? If possible, provide a recipe for reproducing the error.

2. What did you expect to see?

3. What did you see instead?

4. Versions of the cluster

    - DM version (run `dmctl -V` or `dm-worker -V` or `dm-master -V`):

        ```
        DM master until https://github.com/pingcap/dm/commit/0ca7f5424d5bdd877d99b95d5824f08e68ddba0b
        ```

[dm-worker-2_default_dm-worker-97ea66f4c3f18b498b71eb9299d8223c00c8bf6cf4d3f4396691cec79fa6a089.log](https://github.com/pingcap/dm/files/5363382/dm-worker-2_default_dm-worker-97ea66f4c3f18b498b71eb9299d8223c00c8bf6cf4d3f4396691cec79fa6a089.log)

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve? 
close #1143 

### What is changed and how it works?
add `RWMutex` for `RemoteCheckPoint.restoringFiles`

### Check List 

Tests 

 - pass original test

Code changes

Side effects

 - Possible performance regression (@GMHDBJD )

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

**Comment by lance6716** [comment]
/run-all-tests

seems a deadlock caused by dumpling test and v1upgrade test

File: loader/checkpoint.go [hunk 1]
```diff
@@ -416,17 +435,10 @@ func (cp *RemoteCheckPoint) Count(tctx *tcontext.Context) (int, error) {
 }
 
 func (cp *RemoteCheckPoint) String() string {
-	// `String` is often used to log something, it's not a big problem even fail,
-	// so 1min should be enough.
-	tctx2, cancel := cp.logCtx.WithTimeout(time.Minute)
-	defer cancel()
-
-	if err := cp.Load(tctx2); err != nil {
```
Comment (csuzhangxc) [inline loader/checkpoint.go:424]: Do not need to `Load` even for the first time, right?

File: loader/checkpoint.go [hunk 2]
```diff
@@ -205,7 +210,9 @@ func (cp *RemoteCheckPoint) Load(tctx *tcontext.Context) error {
 
 // GetRestoringFileInfo implements CheckPoint.GetRestoringFileInfo
 func (cp *RemoteCheckPoint) GetRestoringFileInfo(db, table string) map[string][]int64 {
-	if tables, ok := cp.restoringFiles[db]; ok {
+	cp.restoringFiles.RLock()
+	defer cp.restoringFiles.RUnlock()
+	if tables, ok := cp.restoringFiles.pos[db]; ok {
```
Comment (csuzhangxc) [inline loader/checkpoint.go:215]: Do we need to return a deep copy (and for `GetAllRestoringFileInfo`)?

File: loader/checkpoint.go (hunk 1 above)
Comment (lance6716) [reply loader/checkpoint.go:424]: I guess checkpoint are not cached in memory before, so this `String()` need to fetch from downstream DB to show latest checkpoint. now it's no need

File: loader/checkpoint.go (hunk 2 above)
Comment (lance6716) [reply loader/checkpoint.go:216]: https://github.com/pingcap/dm/pull/1145/commits/2be83bf8e04ac8a52d0a8b74ee8a9e29b1a57eb7

**Review by csuzhangxc — COMMENTED** [review]
LGTM

**Review by GMHDBJD — APPROVED** [review]
LGTM

**Pull request merged**